### ISSUE TITLE
@overloads and Union hinting lead to different outcome

### ISSUE THREAD

**Issue by kalfa** [python/mypy#14006]
**Bug Report**

As far as I understand the following to ways are equivalent (well, overloading is more precise and that's the reason for which I'd like to use it).

```
def bar(a: Union[int, str]) -> Union[int, str]:
    return {}
```

will report
`error: Incompatible return value type (got "Dict[<nothing>, <nothing>]", expected "Union[int, str]")`

while

```
@overload
def foo(a: int) -> int: ...
@overload
def foo(a: str) -> str: ...
def foo(a):
      return {}
```
does not report anything.

**Expected behaviour**

also overloading reports misalignment from what is declared (either int or str) and what actually the implementation return (dict)

**Comment by hauntsaninja** [comment]
See https://mypy.readthedocs.io/en/stable/more_types.html#type-checking-the-implementation mypy will type check the implementation according to what it is hinted as, which is effectively `foo(a: Any) -> Any` because it's untyped. If you use `--strict`, you'll get an error asking you to add types.
